imap/imap.go

```diff
@@ -150,203 +65,181 @@ func Validate(s *models.IMAP) error {
-	msgCount := 0
-	for _, rsp := range cmd.Data {
-		for _, uid := range rsp.SearchResults() {
-			msgCount++
-			seq.AddNum(uid)
-		}
+	imapclient, err := mbox.newClient()
+	if err != nil {
+		return emails, fmt.Errorf("failed to create IMAP connection: %s", err)
 	}
 
-	if seq.Empty() {
-		return
-	}
+	defer imapclient.Logout()
 
-	fCmd, err := imap.Wait(client.UIDFetch(seq, "INTERNALDATE", "BODY[]", "UID", "RFC822.HEADER"))
+	// Search for unread emails
+	criteria := imap.NewSearchCriteria()
+	criteria.WithoutFlags = []string{"\\Seen"}
+	seqs, err := imapclient.Search(criteria)
 	if err != nil {
-		responses <- Response{Err: fmt.Errorf("unable to perform uid fetch: %s", err)}
-		return
+		return emails, err
 	}
 
-	var email Email
-	for _, msgData := range fCmd.Data {
-		msgFields := msgData.MessageInfo().Attrs
-
-		// make sure is a legit response before we attempt to parse it
-		// deal with unsolicited FETCH responses containing only flags
-		// I'm lookin' at YOU, Gmail!
-		// http://mailman13.u.washington.edu/pipermail/imap-protocol/2014-October/002355.html
-		// http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch
-		if _, ok := msgFields["RFC822.HEADER"]; !ok {
-			continue
-		}
+	if len(seqs) > 0 {
+		seqset := new(imap.SeqSet)
+		seqset.AddNum(seqs...)
+		section := &imap.BodySectionName{}
+		items := []imap.FetchItem{imap.FetchEnvelope, imap.FetchFlags, imap.FetchInternalDate, section.FetchItem()} // Check this
+		messages := make(chan *imap.Message)
 
-		email, err = NewEmail(msgFields)
-		if err != nil {
-			responses <- Response{Err: fmt.Errorf("unable to parse email: %s", err)}
-			return
-		}
+		go func() {
+			if err := imapclient.Fetch(seqset, items, messages); err != nil {
+				log.Error("Error fetching emails: ", err.Error()) // TODO: How to handle this, need to propogate error out
+			}
+		}()
+
+		// Step through each email
+		for msg := range messages {
+			// Extract raw message body. I can't find a better way to do this with the emersion library
+			var em *email.Email
+			var buf []byte
+			for _, value := range msg.Body {
+				buf = make([]byte, value.Len())
+				value.Read(buf)
+				break // There should only ever be one item in this map, but I'm not 100% sure
+			}
 
-		responses <- Response{Email: email}
+			//Remove CR characters, see https://github.com/jordan-wright/email/issues/106
+			tmp := string(buf)
+			re := regexp.MustCompile(`\r`)
+			tmp = re.ReplaceAllString(tmp, "")
+			buf = []byte(tmp)
 
-		if !markAsRead {
-			err = removeSeen(client, imap.AsNumber(msgFields["UID"]))
+			rawBodyStream := bytes.NewReader(buf)
+			em, err = email.NewEmailFromReader(rawBodyStream) // Parse with @jordanwright's library
 			if err != nil {
-				responses <- Response{Err: fmt.Errorf("unable to remove seen flag: %s", err)}
-				return
+				return emails, err
 			}
-		}
 
-		if delete {
-			err = deleteEmail(client, imap.AsNumber(msgFields["UID"]))
+			// Reload the reader 🔫
```

Comment:
Let's avoid having emojis in the comments if possible 😄 Go totally supports it, but I'd like to keep things easy.

Reply:
🙃